### ISSUE TITLE
Out-of-memory for vg construct

### ISSUE THREAD

**Issue by foggieding** [vgteam/vg#2978]
**1. What were you trying to do?**
I am using the vg tools to construct graphs and genotype SVs on multiple samples. While I don' t know the actual running time and memory for each command.
I use known SV list to construct the graph (1.5M SVs in total). I tried various memory settings (10G, 50G, 100G and 500G) and failed.

**2. What did you want to happen?**
I want to know the memory settings and running time for vg construct. Also, for a large list (1.5M) of SVs, can I build graph using this command?

**3. What actually happened?**
I failed to build the graph. Now I am using 1.2T memory to build the graph while the program is still running.

**4. If you got a line like `Stack trace path: /somewhere/on/your/computer/stacktrace.txt`, please copy-paste the contents of that file here:**

```
Place stacktrace here.
```

**5. What data and command can the vg dev team use to make the problem happen?**

**6. What does running `vg version` say?**

```
Place vg version output here
```
vg version v1.26.1 "Stornara"
Compiled with g++ (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0 on Linux
Linked against libstd++ 20191114
Built by anovak@octagon

Looking forward to your reply and thanks a lot!

**Comment by glennhickey** [comment]
It's best to use a separate `vg construct` command for each chromosome if you aren't already.  There's an example here https://github.com/vgteam/vg/wiki/Working-with-a-whole-genome-variation-graph

**Comment by adamnovak** [comment]
The `vg construct` command itself is not supposed to use a lot of memory; it streams through the VCF and streams its output.

Is it possible your large SV collection has SVs that overlap each other to the point that there are giant groups of SVs with no regions of fixed, nonvariable sequence between them? `vg construct` expects to be able to break up its input into smaller problems of one or a few variants that overlap each other. If it ends up trying to work on a lot of variants at once it could probably run out of memory.

Can you share the exact command line you are using? You probably need `--handle-sv` AKA `-S` to get it to even try to do the SVs, and you might want `--flat-alts` AKA `-f` to get it to not try and realign the alts of the SVs to each other most parsimoniously. Also, if you give it `--progress` AKA `-p` it will print out a log of what it is working on, which may help with debugging.

**Comment by foggieding** [comment]
> It's best to use a separate `vg construct` command for each chromosome if you aren't already. There's an example here https://github.com/vgteam/vg/wiki/Working-with-a-whole-genome-variation-graph

Thanks and it' s a great suggestion to run for each chromosome! And I' m working on it and running the command separately.

**Comment by foggieding** [comment]
> The `vg construct` command itself is not supposed to use a lot of memory; it streams through the VCF and streams its output.
> 
> Is it possible your large SV collection has SVs that overlap each other to the point that there are giant groups of SVs with no regions of fixed, nonvariable sequence breaking them up? `vg construct` expects to be able to break up its input into smaller problems of one or a few variants that overlap each other. If it ends up trying to work on a lot of variants at once it could probably run out of memory.
> 
> Can you share the exact command line you are using? You probably need `--handle-sv` AKA `-S` to get it to even try to do the SVs, and you might want `--flat-alts` AKA `-f` to get it to not try and realign the alts of the SVs to each other most parsimoniously, which could be using a lot of memory. Also, if you give it `--progress` AKA `-p` it will print out a log of what it is working on, which may help with debugging.

Yes! I think you are right and actually I' m working on a large SV collection that may overlap each other. Since the SV calls are from various individuals and I haven' t finger out the proper threshold to merge and prune. So I just simply merge them and to do genotype. The exact command is "vg construct -S --flat-alts -R chr${i} -r $refPath -v $sortedVcf -t 1 -p -a" . Do you have any suggestions for the large SV collection with these features? Thanks a lot!

**Comment by ekg** [comment]
We have tried building SV graphs using alternative approaches to merge.

One thing you could try is to build a seqwish graph from the reference and
alleles (and flanking sequences) of the SVs. This can be tricky too. In
principle, it's the best way to go, because the VCF representation can be
extremely complex and redundant. It's just that the precise approach to do
so hasn't really been sorted out.

You could test with a small region. The key thing is to generate the SV
alleles in a FASTA file, then map these to each other and to the reference
using minimap2 or edyeet. You can then induce a graph with seqwish.

Typically, the seqwish graphs have some intense local looping motifs due to
inconsistency in the representation of the alignments. The pggb pipeline
attempts to resolve this, so you could also take your sequence set as input
to that. The graphs resulting from pggb are meant to be easy to use in vg,
GraphAligner and other tools because they don't have these small complex
graph motifs. They will be locally partially ordered.

**Comment by foggieding** [comment]
> We have tried building SV graphs using alternative approaches to merge. One thing you could try is to build a seqwish graph from the reference and alleles (and flanking sequences) of the SVs. This can be tricky too. In principle, it's the best way to go, because the VCF representation can be extremely complex and redundant. It's just that the precise approach to do so hasn't really been sorted out. You could test with a small region. The key thing is to generate the SV alleles in a FASTA file, then map these to each other and to the reference using minimap2 or edyeet. You can then induce a graph with seqwish. Typically, the seqwish graphs have some intense local looping motifs due to inconsistency in the representation of the alignments. The pggb pipeline attempts to resolve this, so you could also take your sequence set as input to that. The graphs resulting from pggb are meant to be easy to use in vg, GraphAligner and other tools because they don't have these small complex graph motifs. They will be locally partially ordered.


Thanks for your reply!  And I will try it.

**Comment by foggieding** [comment]
> to generate the SV alleles in a FASTA file,

Thanks for your reply and I am wondering how to build the graph. Since my SV list is very complicated and overlapped a lot. Can you give me examples of commands to generate the SV alleles in a FASTA file and map with minimap? Thanks a lot!

**Comment by ekg** [comment]
You might be able to use smoothxg to clean up the representation. This is
very experimental though, and the allele paths in the GFA that you give
smoothxg will probably need to be represented with flanking sequence from
the reference to ensure correct alignment.

**Comment by faithokamoto** [comment]
Seems this was answered best we could at the time?
